source/simulator/helper_functions.cc

```diff
@@ -2403,6 +2404,235 @@ namespace aspect
+
+    FEValues<dim> fe_values (*mapping,
+                             finite_element,
+                             quadrature_formula,
+                             update_values |
+                             update_gradients |
+                             update_quadrature_points |
+                             update_JxW_values);
+
+    FEValues<dim> fe_values_adjoint (*mapping,
+                                     finite_element,
+                                     quadrature_formula,
+                                     update_values |
+                                     update_gradients |
+                                     update_quadrature_points |
+                                     update_JxW_values);
+
+    FEFaceValues<dim> fe_face_values (*mapping,
+                                      finite_element,
+                                      quadrature_formula_face,
+                                      update_values |
+                                      update_gradients |
+                                      update_quadrature_points |
+                                      update_JxW_values);
+
+    // Storage for shape function values for the current solution.
+    // Used for constructing the known side of the CBF system.
+    std::vector<double> phi_eta (dofs_per_cell);
+    std::vector<double> phi_rho (dofs_per_cell);
+
+    // Initializing local rhs and mass matrix
+    Vector<double> local_rhs(dofs_per_cell);
+    FullMatrix<double> local_mass_matrix(dofs_per_cell, dofs_per_cell);
+
+    // Loop over all of the surface cells and if one less than h/3 away from
+    // one of the top or bottom boundaries, assemble CBF system for it.
+    typename DoFHandler<dim>::active_cell_iterator
+    cell = dof_handler.begin_active(),
+    endc = dof_handler.end();
+
+    std::vector<types::global_dof_index> local_dof_indices (dofs_per_cell);
+
+    // Get a pointer to the dynamic topography postprocessor.
+    const Postprocess::DynamicTopography<dim> &dynamic_topography =
+      postprocess_manager.template get_matching_postprocessor<Postprocess::DynamicTopography<dim> >();
+
+    // Get the already-computed dynamic topography solution.
+    const LinearAlgebra::BlockVector topo_vector = dynamic_topography.topography_vector();
+    std::vector<double> topo_values( quadrature_formula.size() );
+
+    // TODO: get this from dynamic topography postprocessor
+    const double density_above = 0;
+
+    for (; cell!=endc; ++cell)
+      if (cell->is_locally_owned())
+        {
+          fe_values.reinit (cell);
+          fe_values_adjoint.reinit (cell);
+
+          // Evaluate the material properties and the solution within the cell
+          MaterialModel::MaterialModelInputs<dim> in(fe_values, cell, introspection, solution);
+          MaterialModel::MaterialModelOutputs<dim> out(fe_values.n_quadrature_points, introspection.n_compositional_fields);
+          material_model->evaluate(in, out);
```

Comment:
Here, and below for the face term you may want to include the material model averaging (your assembler includes it as well, because it gets a precomputed material model inputs object):
```
    MaterialModel::MaterialAveraging::average (parameters.material_averaging,
                                               cell,
                                               fe_values.get_quadrature(),
                                               fe_values.get_mapping(),
                                               out);
```

Reply:
I've added this but it doesn't compile and says it can't find a matching function for call to 'average'. Since this is a function in simulator I would have thought that this is already defined. Do have you an idea why it doesn't find this?